storm-client/src/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutor.java

```diff
@@ -0,0 +1,596 @@
+                partitionIds.add(0L);
+                partitionIdsState.put(PARTITION_IDS_KEY, partitionIds);
+            }
+            latestPartitionId = partitionIds.peekLast();
+            latestPartition = cache.getPinned(latestPartitionId);
+        }
+
+        private void initCache() {
+            long size = maxEventsInMemory / MAX_PARTITION_EVENTS;
+            LOG.info("maxEventsInMemory: {}, partition size: {}, number of partitions: {}",
+                maxEventsInMemory, MAX_PARTITION_EVENTS, size);
+            cache = SimpleWindowPartitionCache.<Long, WindowPartition<T>>newBuilder()
+                .maximumSize(size)
+                .removalListener(new RemovalListener<Long, WindowPartition<T>>() {
+                    @Override
+                    public void onRemoval(Long pid, WindowPartition<T> p, RemovalCause removalCause) {
+                        Objects.requireNonNull(pid, "Null partition id");
+                        Objects.requireNonNull(p, "Null window partition");
+                        LOG.debug("onRemoval for id '{}', WindowPartition '{}'", pid, p);
+                        try {
+                            windowPartitionsLock.lock();
+                            if (p.isEmpty() && pid != latestPartitionId) {
+                                // if the empty partition was not invalidated by flush, but evicted from cache
+                                if (removalCause != RemovalCause.EXPLICIT) {
+                                    deletePartition(pid);
+                                    windowPartitionsState.delete(pid);
+                                }
+                            } else if (p.isModified()) {
+                                windowPartitionsState.put(pid, p);
+                            } else {
+                                LOG.debug("WindowPartition '{}' is not modified", pid);
+                            }
+                        } finally {
+                            windowPartitionsLock.unlock();
+                        }
+                    }
+                }).build(new CacheLoader<Long, WindowPartition<T>>() {
+                    @Override
+                   public WindowPartition<T> load(Long id) {
+                        LOG.debug("Load partition: {}", id);
+                        // load from state
+                        try {
+                            windowPartitionsLock.lock();
+                            return windowPartitionsState.get(id, new WindowPartition<>(id));
+                        } finally {
+                            windowPartitionsLock.unlock();
+                        }
+                    }
+                    }
+                );
+        }
+
+        private void deletePartition(long pid) {
+            LOG.debug("Delete partition: {}", pid);
+            try {
+                partitionIdsLock.lock();
+                partitionIds.remove(pid);
+                partitionIdsState.put(PARTITION_IDS_KEY, partitionIds);
```

Comment:
Ok I wrongly assumed the comment was for `initPartitions`, so discard my previous comment. 
Here we first update the `partitionIds` member variable. The put is to save the value into the kv-state. The partitionIdsState state has a single entry something like `"key" -> List of ids`.